Bugfix/issue 150 and 154

wtoorop opened the pull request:
Resolve issue #154 and partially resolve #150 

For reference, given this zone:
```
$ORIGIN example.
$TTL    1
@       SOA     ns admin 1 7200 3600 1209600 3600
        ns      @
        A       192.0.2.1
        TXT     "ab"cd ef"gh" "ij""kl"
test    TXT     (
  "abcdef"
  "ghijk")
```

BIND prints this as:
```
$ named-checkzone -o - example example 
zone example/IN: loaded serial 1
example.	      1	IN SOA		ns.example. admin.example. 1 7200 3600 1209600 3600
example.	      1	IN NS		example.
example.	      1	IN A		192.0.2.1
example.	      1	IN TXT		"ab" "cd" "ef" "gh" "ij" "kl"
test.example.	      1	IN TXT		"abcdef" "ghijk"
OK
```

With this PR, nsd prints this as:
```
$ nsd-checkzone -p example example 
$ORIGIN .
example	1	IN	SOA	ns.example. admin.example. (
		1 7200 3600 1209600 3600 )
	1	IN	NS	example.
	1	IN	A	192.0.2.1
	1	IN	TXT	"ab" "cd" "ef\"gh\"" "ij" "kl"
$ORIGIN example.
test	1	IN	TXT	"abcdef" "ghijk"
; zone example is ok
```

So remaining issue is dotted_str before quoted-string (if we want to align that with bind too).

For reference, ldns-read zone says:
```
$ ldns-read-zone example 
example.	1	IN	SOA	ns.example. admin.example. 1 7200 3600 1209600 3600
example.	1	IN	NS	example.
example.	1	IN	A	192.0.2.1
example.	1	IN	TXT	"ab" "d" "ef\"gh\"" "ij" "kl\""
test.example.	1	IN	TXT	"abcdef" "ghijk"
```

wtoorop committed:
Revert "- Fix #150: TXT record validation difference with BIND."

This reverts commit dea8d2c5bfba4317cb41e179c3ed341e250f50f6.

wtoorop committed:
An unquoted string may follow quoted sting without space

And vice versa, which with bind means separate text elements.

```
$ cat example
$ORIGIN example.
$TTL    1
@       SOA     ns admin 1 7200 3600 1209600 3600
        ns      @
        A       192.0.2.1
        TXT     "ab"cd ef"gh" "ij""kl"
test    TXT     (
  "abcdef"
  "ghijk")
$ named-checkzone -o - example example
zone example/IN: loaded serial 1
example.      1 IN SOA          ns.example. admin.example. 1 7200 3600 1209600 3600
example.      1 IN NS           example.
example.      1 IN A            192.0.2.1
example.      1 IN TXT          "ab" "cd" "ef" "gh" "ij" "kl"
test.example. 1 IN TXT          "abcdef" "ghijk"
OK
```

wtoorop committed:
Print zone with nsd-checkzone -p

```
$ cat >example
$ORIGIN example.
$TTL    1
@       SOA     ns admin 1 7200 3600 1209600 3600
        ns      @
        A       192.0.2.1
        TXT     "ab"cd ef"gh" "ij""kl"
test    TXT     (
  "abcdef"
  "ghijk")
$ nsd-checkzone -p example example
$ORIGIN .
example	1	IN	SOA	ns.example. admin.example. (
		1 7200 3600 1209600 3600 )
	1	IN	NS	example.
	1	IN	A	192.0.2.1
	1	IN	TXT	"ab" "cd" "ef\"gh\"" "ij" "kl"
$ORIGIN example.
test	1	IN	TXT	"abcdef" "ghijk"
; zone example is ok
```

wtoorop committed:
Fix dotted_str usage in rdata's

wcawijngaards reviewed (approved):
Perhaps the remaining issue could be solved by making the unquoted_dotted_str end if a `"` is seen.  It should not happen in dotted strings, I think, so that lexer change could make this parser change also solve the efgh issue?

Code looks good to me, and I like the nsd-checkzone feature to print it.

wcawijngaards commented:
This change is what I am proposing, could this fix the remaining issue?

It disallows the `"` in the charstr for unquoted dotted str and that makes the dotted str end if a quote is seen.  Thus the next parse block is the quoted string.  The escaped quote `\"` is still allowed, to write quotes in strings.

```diff
diff --git a/zlexer.lex b/zlexer.lex
index 2c3a3137..99b82821 100644
--- a/zlexer.lex
+++ b/zlexer.lex
@@ -130,7 +130,7 @@ SPACE   [ \t]
 LETTER  [a-zA-Z]
 NEWLINE [\n\r]
 ZONESTR [^ \t\n\r();.\"\$]|\\.|\\\n
-CHARSTR [^ \t\n\r();.]|\\.|\\\n
+CHARSTR [^ \t\n\r();.\"]|\\.|\\\n
 QUOTE   \"
 DOLLAR  \$
 COMMENT ;
```

wcawijngaards commented:
I tested the change, it works and prints `TXT	"ab" "cd" "ef" "gh" "ij" "kl"`.

wtoorop committed:
Suggestion from Wouter to align parsing with how bind does it

```
$ cat example
$ORIGIN example.
$TTL    1
@       SOA     ns admin 1 7200 3600 1209600 3600
        ns      @
        A       192.0.2.1
        TXT     "ab"cd ef"gh" "ij""kl"
test    TXT     (
  "abcdef"
  "ghijk")
$ named-checkzone -o - example example
zone example/IN: loaded serial 1
example.      1 IN SOA          ns.example. admin.example. 1 7200 3600 1209600 3600
example.      1 IN NS           example.
example.      1 IN A            192.0.2.1
example.      1 IN TXT          "ab" "cd" "ef" "gh" "ij" "kl"
test.example. 1 IN TXT          "abcdef" "ghijk"
OK
$ ./nsd-checkzone -p example example
$ORIGIN .
example	1	IN	SOA	ns.example. admin.example. (
		1 7200 3600 1209600 3600 )
	1	IN	NS	example.
	1	IN	A	192.0.2.1
	1	IN	TXT	"ab" "cd" "ef" "gh" "ij" "kl"
$ORIGIN example.
test	1	IN	TXT	"abcdef" "ghijk"
; zone example is ok
```

wtoorop committed:
Align rr-test.tdir with new (bind compatible) text parsing
